### ISSUE TITLE
UseOnlyWhitelistedNamespaces: Allow all except root

### ISSUE THREAD

**Issue by Soullivaneuh** [slevomat/coding-standard#245]
I would like to allows use statement for any namespace, except root.

A `allowUseFromRootNamespace` option is available under the `SlevomatCodingStandard.Namespaces.UseOnlyWhitelistedNamespaces` sniff.

It might feeds my needs, but using this sniffer force me to set a namespaces white-list, that I don't want.

Maybe this sniff should be splitted?

**Comment by kukulich** [comment]
See `ReferenceUsedNamesOnlySniff`

**Comment by Soullivaneuh** [comment]
@kukulich May you please elaborate? I'm using this sniff, but I don't see which option would help me for that.

Here is my configuration:

```xml
<?xml version="1.0"?>
<ruleset name="FlintCI Standard">
    <file>src</file>
    <file>src-dev</file>
    <file>tests</file>

    <exclude-pattern>tests/fixtures/repos</exclude-pattern>

    <rule ref="Symfony">
        <exclude name="Symfony.Commenting.License"/>
        <exclude name="Symfony.Commenting.Annotations.Invalid"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/120 -->
        <exclude name="Symfony.Commenting.FunctionComment"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/119 -->
        <exclude name="Symfony.Functions.ReturnType.Invalid"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/27 -->
        <exclude name="PEAR.Functions.FunctionCallSignature"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/44 -->
        <exclude name="Squiz.Strings.ConcatenationSpacing"/>
        <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/122 -->
        <exclude name="Squiz.Functions.MultiLineFunctionDeclaration.ContentAfterBrace"/>
        <exclude name="Squiz.WhiteSpace.ScopeClosingBrace.ContentBefore"/>
        <exclude name="PEAR.WhiteSpace.ScopeClosingBrace.Line"/>
        <exclude name="Symfony.Formatting.BlankLineBeforeReturn.Invalid"/>
    </rule>

    <!-- @see https://github.com/djoos/Symfony-coding-standard/issues/121 -->
    <rule ref="Symfony.NamingConventions.ValidClassName.InvalidAbstractName">
        <exclude-pattern>*TestCase.php</exclude-pattern>
    </rule>

    <rule ref="SlevomatCodingStandard">
        <exclude name="SlevomatCodingStandard.ControlStructures.DisallowYodaComparison"/>
        <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedClassNameInAnnotation"/>
        <!-- @see https://github.com/slevomat/coding-standard/issues/245 -->
        <exclude name="SlevomatCodingStandard.Namespaces.UseOnlyWhitelistedNamespaces"/>
        <!-- @see https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3471 -->
        <exclude name="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly.ReferenceViaFullyQualifiedNameWithoutNamespace"/>
        <!-- @see https://github.com/slevomat/coding-standard/issues/246 -->
        <exclude name="SlevomatCodingStandard.Namespaces.FullyQualifiedExceptions.NonFullyQualifiedException"/>
    </rule>
    <rule ref="SlevomatCodingStandard.Files.TypeNameMatchesFileName">
        <properties>
            <property
                name="rootNamespaces"
                type="array"
                value="
                    tests=>Tests,
                    src-dev=>Dev,
                "
            />
            <property name="ignoredNamespaces" type="array" value="DoctrineMigrations"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly">
        <properties>
            <property name="allowFullyQualifiedGlobalClasses" value="true"/>
            <property name="allowFullyQualifiedGlobalFunctions" value="true"/>
            <property name="allowFullyQualifiedGlobalConstants" value="true"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Namespaces.UnusedUses">
        <properties>
            <property name="searchAnnotations" value="true"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.TypeHints.DeclareStrictTypes">
        <properties>
            <property name="newlinesCountBetweenOpenTagAndDeclare" value="2"/>
            <property name="spacesCountAroundEqualsSign" value="0"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.TypeHints.TypeHintDeclaration">
        <properties>
            <property
                name="usefulAnnotations"
                type="array"
                value="
                    @after,
                    @before,
                    @deprecated,
                    @required,
                    @ORM\,
                    @Route,
                    @Method,
                    @IsGranted
                "
            />
            <property name="enableEachParameterAndReturnInspection" value="true"/>
            <property name="enableObjectTypeHint" value="false"/>
        </properties>
    </rule>
    <rule ref="SlevomatCodingStandard.Types.EmptyLinesAroundTypeBraces">
        <properties>
            <property name="linesCountAfterOpeningBrace" value="0"/>
            <property name="linesCountBeforeClosingBrace" value="0"/>
        </properties>
    </rule>
</ruleset>
```

So yes, I use `allowFullyQualifiedGlobalClasses`. But it does not enforce to not use `use DateTime` for example.

Thanks.

**Comment by kukulich** [comment]
You’re right, it´s not forced but it´s the best you can do with current version. You can also prepare PR with “FullyQualifiedGlobalClassesSniff” based on FullyQualifiedGlobalFunctionsSniff.

**Comment by Soullivaneuh** [comment]
Yes, in this case, could you please re-open this issue?

I don't have enough time right now, but maybe somebody else could be interested to make a PR. :wink:

**Comment by Soullivaneuh** [comment]
BTW, on which sniffer should we work on to make this feasible?
